toolchain/parse/handle_import_and_package.cpp

```diff
@@ -5,113 +5,168 @@
+                                   NodeKind directive,
+                                   llvm::function_ref<void()> on_parse_error)
+    -> void {
+  Tree::PackagingNames names{.node = Node(state.subtree_start)};
+  if (directive != NodeKind::LibraryDirective) {
+    if (auto package_name_token =
+            context.ConsumeIf(Lex::TokenKind::Identifier)) {
+      names.package_id = context.tokens().GetIdentifier(*package_name_token);
+      context.AddLeafNode(NodeKind::PackageName, *package_name_token);
+    } else if (directive == NodeKind::PackageDirective ||
+               !context.PositionIs(Lex::TokenKind::Library)) {
+      CARBON_DIAGNOSTIC(ExpectedIdentifierAfterPackage, Error,
+                        "Expected identifier after `package`.");
+      CARBON_DIAGNOSTIC(ExpectedIdentifierAfterImport, Error,
+                        "Expected identifier or `library` after `import`.");
+      context.emitter().Emit(*context.position(),
+                             directive == NodeKind::PackageDirective
+                                 ? ExpectedIdentifierAfterPackage
+                                 : ExpectedIdentifierAfterImport);
+      on_parse_error();
+      return;
+    }
   }
 
-  Tree::ApiOrImpl api_or_impl;
-  if (is_package) {
-    switch (next_kind) {
-      case Lex::TokenKind::Api: {
-        context.AddLeafNode(NodeKind::PackageApi, context.Consume());
-        api_or_impl = Tree::ApiOrImpl::Api;
-        break;
-      }
-      case Lex::TokenKind::Impl: {
-        context.AddLeafNode(NodeKind::PackageImpl, context.Consume());
-        api_or_impl = Tree::ApiOrImpl::Impl;
-        break;
+  // Parse the optional library keyword.
+  bool accept_default = !names.package_id.is_valid();
+  if (directive == NodeKind::LibraryDirective) {
+    auto library_id = HandleLibraryName(context, accept_default);
+    if (!library_id) {
+      on_parse_error();
+      return;
+    }
+    names.library_id = *library_id;
+  } else {
+    auto next_kind = context.PositionKind();
+    if (next_kind == Lex::TokenKind::Library) {
+      auto library_token = context.ConsumeChecked(Lex::TokenKind::Library);
+      auto library_subtree_start = context.tree().size();
+      auto library_id = HandleLibraryName(context, accept_default);
+      if (!library_id) {
+        on_parse_error();
+        return;
       }
-      default: {
-        CARBON_DIAGNOSTIC(ExpectedApiOrImpl, Error,
-                          "Expected `api` or `impl`.");
-        context.emitter().Emit(*context.position(), ExpectedApiOrImpl);
-        ExitOnParseError(context, state, directive);
+      names.library_id = *library_id;
+      context.AddNode(NodeKind::LibrarySpecifier, library_token,
+                      library_subtree_start,
+                      /*has_error=*/false);
+    } else {
+      auto next_kind = context.PositionKind();
+      if (next_kind == Lex::TokenKind::StringLiteral ||
+          (accept_default && next_kind == Lex::TokenKind::Default)) {
```

Comment:
I think this can just be:

```suggestion
    } else if (next_kind == Lex::TokenKind::StringLiteral ||
               (accept_default && next_kind == Lex::TokenKind::Default)) {
```

But I'm finding the conditions a bit hard to track here... Not sure I see anything substantially better you can do though.

Reply:
Oops, fixed the redundant next_kind.

Yeah though -- the best I was coming up with was trying to simplify the syntax. ;)